### PR TITLE
Mobile 3099

### PR DESCRIPTION
**Pull request by crazyserver**

### CODE REVIEW

File: src/core/features/course/pages/module-preview/module-preview.page.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+     * Refresh the data.
+     *
+     * @param refresher Refresher.
+     * @return Promise resolved when done.
+     */
+    async doRefresh(refresher?: IonRefresher): Promise<void> {
+
+        await CoreCourse.invalidateModule(this.module.id);
+
+        this.fetchModule();
```
Comment (dpalou) [inline src/core/features/course/pages/module-preview/module-preview.page.ts:99]: Shouldn't you do fetchModule(true)?

File: src/core/features/course/components/module/core-course-module.html [hunk 2]
```diff
@@ -75,9 +75,9 @@
                 </div>
 
                 <!-- Module completion. -->
-                <core-course-module-completion *ngIf="module.completiondata" [completion]="module.completiondata" [moduleName]="module.name"
-                    [moduleId]="module.id" [showCompletionConditions]="showCompletionConditions"
-                    [showManualCompletion]="showManualCompletion" (completionChanged)="completionChanged.emit($event)">
+                <core-course-module-completion *ngIf="module.completiondata && module.uservisible" [completion]="module.completiondata"
+                    [moduleName]="module.name" [moduleId]="module.id" [showCompletionConditions]="showCompletionConditions"
+                    [showManualCompletion]="true" (completionChanged)="completionChanged.emit($event)">
```
Comment (dpalou) [inline src/core/features/course/components/module/core-course-module.html:80]: Why did you change showManualCompletion to true? IIRC it depends on the module type.

**Review by dpalou — CHANGES_REQUESTED** [review]

File: src/addons/mod/forum/components/index/index.html [hunk 3]
```diff
@@ -140,8 +140,11 @@
         </ng-container>
     </core-loading>
 
-    <ion-fab slot="fixed" core-fab vertical="bottom" horizontal="end" *ngIf="forum && canAddDiscussion">
-        <ion-fab-button (click)="openNewDiscussion()" [attr.aria-label]="addDiscussionText">
+    <core-course-module-navigation [hidden]="!loaded" [courseId]="courseId" [currentModuleId]="module.id">
+    </core-course-module-navigation>
+
+    <ion-fab slot="fixed" core-fab vertical="bottom" horizontal="end">
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/addons/mod/forum/components/index/index.html:146]: Thanks!

File: src/core/features/course/components/module-navigation/module-navigation.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                return false;
+            }
+
+            currentModuleIndex = section.modules.findIndex((module: CoreCourseModule) => module.id == this.currentModuleId);
+
+            return currentModuleIndex >= 0;
+        });
+
+        if (currentSectionIndex < 0) {
+            // Nothing found. Return.
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/course/components/module-navigation/module-navigation.ts:188]: Nice cathc, thanks

File: src/addons/mod/glossary/components/index/addon-mod-glossary-index.html [hunk 5]
```diff
@@ -96,8 +96,11 @@ <h2>{{ getDivider!(entry) }}</h2>
         </core-infinite-loading>
     </core-loading>
 
-    <ion-fab slot="fixed" core-fab vertical="bottom" horizontal="end" *ngIf="canAdd">
-        <ion-fab-button (click)="openNewEntry()" [attr.aria-label]="'addon.mod_glossary.addentry' | translate">
+    <core-course-module-navigation [hidden]="!loaded" [courseId]="courseId" [currentModuleId]="module.id">
+    </core-course-module-navigation>
+
+    <ion-fab slot="fixed" core-fab vertical="bottom" horizontal="end">
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/addons/mod/glossary/components/index/addon-mod-glossary-index.html:102]: Done

File: src/core/features/course/components/module-navigation/module-navigation.ts [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    }
+
+    /**
+     * Section is visible by the user and its not stealth
+     *
+     * @param section Section to check.
+     * @return Wether the module is available to the user or not.
+     */
+    protected isSectionAvailable(section: CoreCourseWSSection): boolean {
+        return !(section.uservisible === false) && section.id != CoreCourseProvider.STEALTH_MODULES_SECTION_ID;
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/course/components/module-navigation/module-navigation.ts:257]: There was something inside the () before, thanks for noticing!

File: src/core/components/navigation-bar/navigation-bar.ts [hunk 7]
```diff
@@ -12,48 +12,102 @@
...
 })
-export class CoreNavigationBarComponent {
-
-    @Input() previous?: unknown; // Previous item. If not defined, the previous arrow won't be shown.
-    @Input() previousTitle?: string; // Previous item title. If not defined, only the arrow will be shown.
-    @Input() next?: unknown; // Next item. If not defined, the next arrow won't be shown.
-    @Input() nextTitle?: string; // Next item title. If not defined, only the arrow will be shown.
-    @Input() info = ''; // Info to show when clicking the info button. If not defined, the info button won't be shown.
-    @Input() title = ''; // Title to show when seeing the info (new page).
+export class CoreNavigationBarComponent implements OnChanges {
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/components/navigation-bar/navigation-bar.ts:32]: Done!

File: src/core/features/course/pages/module-preview/module-preview.page.ts (hunk 1 above)
Comment (crazyserver) [reply src/core/features/course/pages/module-preview/module-preview.page.ts:99]: true!

File: src/addons/mod/h5pactivity/pages/index/index.ts [hunk 8]
```diff
@@ -36,7 +36,8 @@ export class AddonModH5PActivityIndexPage extends CoreCourseModuleMainActivityPa
      * @inheritdoc
      */
     async canLeave(): Promise<boolean> {
-        if (!this.activityComponent || !this.activityComponent.playing || this.activityComponent.isOpeningPage) {
+        if (!this.activityComponent || !this.activityComponent.playing || this.activityComponent.isOpeningPage ||
+                this.activityComponent.moduleNavigation?.isOpening) {
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/addons/mod/h5pactivity/pages/index/index.ts:40]: Great! Removed

File: src/core/components/navigation-bar/navigation-bar.ts [hunk 9]
```diff
@@ -12,48 +12,102 @@
...
+    currentIndex = 0;
+
+    // Function to call when arrow is clicked. Will receive as a param the item to load.
+    @Output() action: EventEmitter<unknown> = new EventEmitter<unknown>();
+
+    /**
+     * @inheritdoc
+     */
+    ngOnChanges(): void {
+        if (!this.items.length) {
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/components/navigation-bar/navigation-bar.ts:57]: Nice catch! Done!

File: src/core/features/course/components/module/core-course-module.html (hunk 2 above)
Comment (crazyserver) [reply src/core/features/course/components/module/core-course-module.html:80]: Oops, now it's recovered. Also in course-module-info were it was previously removed. I leave it true as default.

**Review by dpalou — APPROVED** [review]
